### PR TITLE
🚀 🚀 🚀  amp-carousel ssr: allow render without JS 🚀 🚀 🚀 

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Server-rendering AMP Components should allow us to have UI visible without blocking on JS executaion.
Via disabling JS locally, I was surprised to see that this wasn't the case yet.

We still have two issues stopping `<amp-carousel>` from rendering without JS:
1. Loading extension CSS is still bundled with JS execution. We need to add a `<style>` link to the CSS to the `<head>`.
2. The application of class names which unhide the amp-carousel take place in `layoutCallback` instead of `buildDom`.  In this case it is the addition of `i-amphtml-element` and `i-amphtml-slide-item-show`.

I solve (2) in this PR, and am working on (1) separately.

🚀 🚀 🚀  **results** 🚀 🚀 🚀 
For webpages with their LCP Element in the amp-carouse, we can expect a ~40% improvement to LCP. [source](https://webpagetest.org/video/compare.php?tests=220211_AiDc37_DBC,220211_BiDcYW_BQN)

lcp-improvement

### CODE REVIEW

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 1]
```diff
@@ -939,7 +940,7 @@ export class AmpSlideScroll extends AMP.BaseElement {
         }
         dev()
           .assertElement(this.slideWrappers_[i])
-          .classList.remove(SHOWN_CSS_CLASS);
+          .classList.remove(ClassNames.SLIDES_ITEM_SHOW);
```
Comment (rcebulko) [inline extensions/amp-carousel/0.1/slidescroll.js:943]: This `dev().assertElement` may as well be removed/replaced with `?.`, particularly given the predicate on line 932 already accesses it without any sort of check `this.slideWrappers_[i].classList`

**Review by rcebulko — COMMENTED** [review]
LGTM so far w/ 1 minor tweak but in draft so will hold off approving

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 1 above)
Comment (samouri) [reply extensions/amp-carousel/0.1/slidescroll.js]: Good call! Done

**Comment by samouri** [comment]
**for the reviewer**
There are three notable changes in this PR:

1. All SSRed elements will now have the class `i-amphtml-element`
2. All slidescroll `<amp-carousel>`s will have their first slide start out visible
3. `setStyle` now exclusively uses `setProperty` method instead of ever falling back to assignment on `style` keys. This is accomplished with a new function, `camelCaseToHyphenCase`

File: extensions/amp-carousel/0.1/build-dom.js [hunk 2]
```diff
@@ -229,6 +231,8 @@ function buildSlideScrollCarousel(element) {
     slidesContainer.appendChild(slideWrapper);
     slideWrappers.push(slideWrapper);
   });
+  // Initialize the first slide to be shown.
+  slideWrappers[0]?.classList.add(ClassNames.SLIDES_ITEM_SHOW);
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/build-dom.js:235]: Nit: Move this into the `forEach` above? Also, why isn't it a `map`…

File: src/core/dom/style.js [hunk 3]
```diff
@@ -126,11 +125,7 @@ export function setStyle(element, property, value, opt_units, opt_bypassCache) {
     return;
   }
   const styleValue = opt_units ? value + opt_units : value;
-  if (isVar(propertyName) || SHOULD_USE_SET_PROPERTY.has(propertyName)) {
-    element.style.setProperty(propertyName, styleValue);
-  } else {
-    /** @type {*} */ (element.style)[propertyName] = styleValue;
-  }
+  element.style.setProperty(camelCaseToHyphenCase(propertyName), styleValue);
```
Comment (jridgewell) [inline src/core/dom/style.js:128]: Let's fix `setStylesImportant` while we're here?

File: src/core/dom/style.js [hunk 4]
```diff
@@ -27,6 +27,14 @@ export function camelCaseToTitleCase(camelCase) {
   return camelCase.charAt(0).toUpperCase() + camelCase.slice(1);
 }
 
+/**
+ * @param {string} camelCase camel cased string
+ * @return {string} hyphen-cased string
+ */
+export function camelCaseToHyphenCase(camelCase) {
+  return camelCase.replace(/[A-Z]/g, (match) => '-' + match.toLowerCase());
```
Comment (jridgewell) [inline src/core/dom/style.js:35]: `webkitFoo` can be lowercase, but still needs a leading dash.

**Review by jridgewell — CHANGES_REQUESTED** [review]

File: extensions/amp-carousel/0.1/build-dom.js (hunk 2 above)
Comment (samouri) [reply extensions/amp-carousel/0.1/build-dom.js:235]: I'm confused. Why should this be part of the forEach and what should be a map?

File: src/core/dom/style.js (hunk 4 above)
Comment (samouri) [reply src/core/dom/style.js:35]: This seems tricky

Comment (samouri) [reply src/core/dom/style.js]: Got it

**Comment by samouri** [comment]
There seem to be a long-tail of test cases that create a fake `style` which of course now needs to have a `setProperty` method on it. Hopefully caught them all.

File: extensions/amp-carousel/0.1/build-dom.js (hunk 2 above)
Comment (samouri) [reply extensions/amp-carousel/0.1/build-dom.js:235]: Oh I see, instead of arr.forEach it could be arr.map.
Mind if I punt on that to a small PR tomorrow?

**Comment by samouri** [comment]
@jridgewell: ready for a rereview

**Review by jridgewell — APPROVED** [review]

**Pull request merged**